sdk/core/azure-core/azure/core/pipeline/base_async.py

```diff
@@ -62,22 +62,30 @@ def __init__(self, policy: SansIOHTTPPolicy) -> None:
         super(_SansIOAsyncHTTPPolicyRunner, self).__init__()
         self._policy = policy
 
-    async def send(self, request: PipelineRequest):
+    async def send(self, request: PipelineRequest) -> PipelineResponse:
         """Modifies the request and sends to the next policy in the chain.
 
         :param request: The PipelineRequest object.
         :type request: ~azure.core.pipeline.PipelineRequest
         :return: The PipelineResponse object.
         :rtype: ~azure.core.pipeline.PipelineResponse
         """
-        self._policy.on_request(request)
+        # type ignore on await: https://github.com/python/mypy/issues/7587
+        request_result = self._policy.on_request(request)
+        if hasattr(request_result, '__await__'):
+            await request_result  # type: ignore
```

Comment:
I'm wondering if this can be refactored into the method:
```
async def _await_result(self, func, request):
    result = func(request)
    if hasattr(result, '__await__):
        return await result
    return result
```
Then in the sync runner - maybe we want to check if someone tried to use an async policy?
```
def _await_result(self, func, request):
    result = func(request)
    if hasattr(result, '__await__):
        raise TypeError("Policy {} retuned awaitable object in non-async pipeline.")
    return result
```

Reply:
Added!